### DISCUSSION TITLE
How are Bisq's BTC mining fees for trades calculated?

### DISCUSSION THREAD

**Discussion by pazza83** [bisq-network/bisq#5459] [Q&A]
Thought I would post this question here as I have not seen the specifics of how mining fees on Bisq are calculated anywhere.

From regularly monitoring https://mempool.space/ when looking to place a trade I understand the basics: 

- Quiet mempool generally means lower fees
- Quiet mempool with recently completed block with high average fees = high fees 
- Quiet mempool with current block with high average fees = high fees
- Quiet mempool with current blocks <10 sat/ vB = minimum 10 sats / vB charged

A rough rule of thumb seems to be:

(Last block average s/vB + current block average s/vB) / 2 = current s/vB fees for Bisq's BTC mining fees for trades

On different instances of Bisq prices seem to update at different times. When two instances of Bisq are running then prices are often different. I assume this is due to the nodes they are connected to?

So, questions:

- Is the calculation of how Bisq mining fees documented in Github anywhere?
- Is the minimum fee level of  current Bisq trades 10 sats/vB documented in GitHub issue anywhere?
- At what point are the fees locked-in. I have all the pop-ups on by default and am shown what the mining fees will be before confirming. I need to press confirm twice. Are the fees locked in at the point of the first confirmation or are they only an estimation?

Maybe I am getting too excited by the current low fees but I would love see the 10 sat/vB reversed to save some sats on my trades. I am happy to take the risk that they take longer to confirm when making an offer. I am not time sensitive and happy to wait for the low sats :)

**Comment by huey735** [comment]
I'm pretty sure [this is the code responsible for it](https://github.com/bisq-network/bisq/blob/master/pricenode/src/main/java/bisq/price/mining/providers/MempoolFeeRateProvider.java).

**Comment by ghost** [comment]
> Is the calculation of how Bisq mining fees documented in Github anywhere?

Mining fee rate is calculated by a 3rd party.  https://mempool.space/api/v1/fees/recommended
https://github.com/bisq-network/projects/issues/27

> Is the minimum fee level of current Bisq trades 10 sats/vB documented in GitHub issue anywhere?

I think it is not documented in an issue.  @wiz frequently explains the reasoning behind having 10 sats/vB minimum: "_...because of mining variance and random people suddenly broadcasting huge amounts of transactions_".   https://github.com/bisq-network/bisq/issues/4575

Looking at the code, the minimum 10 sats/vB limit has been in the pricenode as far back as its history goes (since 2018).  I would point you to @cbeams and @wiz if you want more history.


> On different instances of Bisq prices seem to update at different times. When two instances of Bisq are running then prices are often different. I assume this is due to the nodes they are connected to?

When idle, Bisq polls the pricenode for fee details at 5 minute intervals.  When making an offer, if the fee details have not been polled within the last 2 minutes, they are refreshed.


> At what point are the fees locked-in. I have all the pop-ups on by default and am shown what the mining fees will be before confirming. I need to press confirm twice. Are the fees locked in at the point of the first confirmation or are they only an estimation?

They are locked-in at the first pop-up which displays the fees AFAIK.

---

I think your suggestion of having low priority draft offers queued-up for entry is a good idea.  But I don't think it is worthwhile to worry about using fees below 10 sats/vB.  Back a year or two ago when the mempool would completely empty every night, maybe those days are gone now?

For ultra low fees the alternative security models of Misq / Sisq seem more promising.  Maybe best to keep the high security model of on-chain Bitcoin the way it is already without introducing more support and development risk.

**Comment by pazza83** [comment]
Thanks @jmacxx 

I do not think the Bisq trade fees come, at least directly, from https://mempool.space/api/v1/fees/recommended 

These seem to be be the fees displayed on mempool.space 

If the fees where this it would be great. But the fees mostly higher, assuming it has something to do with an algorithm used to calculated the price and/or average values of some of the seed nodes.

Thanks for letting me know that 10 sats v/B minimum is historic, and when the mining fees are locked-in.

@huey735 

Thanks for this. From the code it show:

` .map(r -> Math.multiplyExact(r, 2)) // multiply the minimumFee by 2 (per wiz)`

Not sure what this means exactly, but appears in some circumstances the min fee shown on mempool.space is multiplied by 2 to calculate Bisq's BTC mining fees for trading. 

Are you able to shed any light on this @wiz?
